Parser csv

Gispoo opened the pull request.

Gispoo committed:
Написал CSV парсер. пытался соблюдать СОЛИД.

Gispoo committed:
поправил имена, вывод и добавил тесты для считателя

Gispoo committed:
Написал тесты для ридера и сортера (пришлось поправить ридера)

Gispoo committed:
Написал тесты для врайтера

Gispoo committed:
изменил имя организации и example)
добавил в гит игнор лишние файлы

Pr0meTheuSS commented on tasks/ParserCSV/src/test/resources/output1.txt:
```diff
@@ -0,0 +1,102 @@
+и	8.0	0.06557377%
```
Пристаю по мелочам, но если в колонке с процентами сложить все чиселки, то 100% не получается (либо % лишний, либо запятую на два знака перенести).

Pr0meTheuSS commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/Reader.java:
```diff
@@ -0,0 +1,50 @@
+    public Reader(InputStream inputStream) {
+        try {
+            final Scanner scanner = new Scanner(inputStream);
+            System.out.print("Enter path to input file: ");
```
Не есть хорошо, что в классе, который должен только лишь читать что-то из потока, происходит запрос пути к пользователю через Scanner

Pr0meTheuSS commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvWriter.java:
```diff
@@ -0,0 +1,50 @@
+    public CsvWriter() {
+        try {
+            final Scanner scanner = new Scanner(System.in);
+            System.out.print("Enter path to output file: ");
```
Тоже непонятно, почему класс для вывода не может поток или filePath получить из вне, а не запрашивать его через консоль у пользователя.

Pr0meTheuSS commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvProcessor.java:
```diff
@@ -0,0 +1,23 @@
+        Map<String, Integer> wordFrequencies = new HashMap<>();
+        Reader reader = new Reader();
+        for (String line; (line = reader.getLine()) != null; ) {
+            String[] words = line.split(" ");
```
Не уверен, что будет работать с 2+ пробелами подряд.
Предложил бы использовать регулярное выражение.

Pr0meTheuSS commented on tasks/ParserCSV/src/test/java/ReaderTest.java:
```diff
@@ -0,0 +1,61 @@
+        InputStream inputStream = new ByteArrayInputStream(testFilePath.getBytes());
+
+        Reader reader = new Reader(inputStream);
+        assertNotNull(reader);
```
А он вообще может быть нулл, после того как мы new Reader сделали?

Gispoo committed:
Ридер и врайтер больше не запрашивают ничего из консоли.
Добавил класс, который спрашивает пользователя пути до нужных файлов и возвращать потоки к ним.
Удалил из индекса гита лишнее.

Gispoo committed:
закрыл Scanner из GetterStream

Gispoo committed:
Добавил метод обработки данных и вывода в CsvParser.
Поправил гитигнор

Gispoo committed:
добавил регулярное выражение

Pikau12 commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/Reader.java:
```diff
@@ -0,0 +1,30 @@
+        }
+    }
+
+    public void close() {
```
Просто предложение:
Можно сделать так, чтобы блок try/catch сам закрывал ридер, а не ты ручками.

VMenshchikov replied:
хорошее замечание. так же при наличии такой функции стоит имплементировать интерфейсы Clasable и Autoclosable

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvProcessor.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.HashMap;
+
+public class CsvProcessor {
+    private GetterStream filePaths;
```
Не совсем понятно, что такое GetterStream. Если поток слов, назови WordsStream

sergomyaso replied:
А это не поток слов, это поток аругментов. Точно нет, в аругментах конструктора нужно конфиг ну и неймнг явно поменяй, сначала подумал, что ты вообще о другом

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvProcessor.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Map;
+import java.util.HashMap;
+
+public class CsvProcessor {
```
А где конструктор? Ты будешь на каждый вызов метода processCsv вызывать? Не очень удачная идея, потому что у тебя нет возможности кофигурировать твой парсер. Плюс нарушаешь правила солид, про инверсию зависимостей и про то, что у тебя зависимости создаются внутри, а не прокидываются. 

1. Заведи интерфейсы
2. Заведи имплементации
3. Создай класс контейнер, внутри создай все зависимости
4. В классе Main создай контейнер (всю логику унеси внутрь класса, чтобы вы мэйне было только Container cnt = new Container())
5. В идеале сделай FactoryMethod для создания каждой зависимости с методом create() а внутрь всякие там аругменты
6. Передай в конструктор зависимости из контейнера (пихай через гетеры getWordStream() и так далеее)

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvProcessor.java:
```diff
@@ -0,0 +1,44 @@
+        wordFrequencies = new HashMap<>();
+        Reader fileReader = new Reader(filePaths.getInputFile());
+
+        for (String currentLine; (currentLine = fileReader.readLine()) != null; ) {
```
Сделай StringBuilder вне цикла
а в форе currentLine = fireader,ReadLine(); currentLIne != null

VMenshchikov replied:
исправлять не стоит, но можешь спросить у одногруппников(если они не просто скопировали из нейросетей, а еще узнали что это) про java.util.Stream и как можно использовать его для более эффективной обработки файла через bufferedReader.lines()

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvProcessor.java:
```diff
@@ -0,0 +1,44 @@
+
+        for (String currentLine; (currentLine = fileReader.readLine()) != null; ) {
+            String[] wordsInLine = currentLine.split("\\s+");
+            wordCounter.CalculateFrequencyWords(wordFrequencies, wordsInLine);
```
Ты в цикле считаешь свой калькулейт, значит AddWordsFrequency. Атактипа ты каждый раз с нуля считаешЬ, а это не так, ты икриментишь частотут в конутере

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvProcessor.java:
```diff
@@ -0,0 +1,44 @@
+        fileReader.close();
+    }
+
+    private void outputData() {
```
Гла-гол, outputData может быть объект, но не функа. ФУнка всегда глагол.

sergomyaso replied:
Где javadoc? Нааааа всееее джава док

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvProcessor.java:
```diff
@@ -0,0 +1,44 @@
+
+    private void outputData() {
+        CsvWriter csvWriter = new CsvWriter(filePaths.getOutputFile());
+        filePaths.close();
```
А если развалит, исключения будут?

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvWriter.java:
```diff
@@ -0,0 +1,41 @@
+
+    public void Write(List<Map.Entry<String, Integer>> entries, int numWords) {
+        if (writer == null) {
+            System.err.println("Не удалось создать файл для записи. Данные не будут сохранены.");
```
А исключения не завезли?

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/GetterStream.java:
```diff
@@ -0,0 +1,65 @@
+     * Получает InputStream для чтения данных из файла, путь к которому вводится через консоль.
+     * Обеспечивает обработку исключений и повторный запрос пути при ошибке.
+     * @return InputStream для чтения из файла, или null, если не удалось получить InputStream.
+     */
```
Ну где-то же ты сделал, делай везде!))

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/GetterStream.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.Scanner;
+import java.io.FileNotFoundException;
+
+public class GetterStream {
```
А класс то чего не статичный, сделай его статичным классом, состояния то у него нет

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvWriter.java:
```diff
@@ -0,0 +1,41 @@
+            for (Map.Entry<String, Integer> entry : entries) {
+                String word = entry.getKey();
+                float freq = entry.getValue();
+
```
Го в CSVFormatter вынесем?

VMenshchikov commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/CsvProcessor.java:
```diff
@@ -0,0 +1,44 @@
+        Reader fileReader = new Reader(filePaths.getInputFile());
+
+        for (String currentLine; (currentLine = fileReader.readLine()) != null; ) {
+            String[] wordsInLine = currentLine.split("\\s+");
```
если я правильно помню условия, то любой символ, кроме цифр и букв, является разделителями, те ! @ # $ % ^ & * ( ) _ - + и все остальное являются разделителями, а не частями слов

VMenshchikov commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/Reader.java:
```diff
@@ -0,0 +1,30 @@
+        try {
+            return reader.readLine();
+        } catch (IOException e) {
+            System.err.println("Ошибка при чтении файла: " + e.getMessage());
```
при исключении не закрыл файл. критично

VMenshchikov commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/Reader.java:
```diff
@@ -0,0 +1,30 @@
+import java.io.FileReader;
+import java.io.IOException;
+
+public class Reader {
```
указав Closable ты получишь возможность передавать это в другие классы/методы. Указав Autoclosable ты получишь замечательную возможность обернуть код в соседнем классе в **try with resources** и забыть думать об утечке.

пример может быть не корректным:
помни, что класс, который ты написал, будешь использовать не только ты
если бы мне понадобилось распарсить пачку файлов(допустим из архива с парой томов одной книги), то мое первое желание было бы засунуть пачку Reader`ов в java.util.stream.Stream, но случилась бы утечка, тк не реализован интерфейс 
мораль: метод close() хорошо, а интерфейс намного лучше. Переписывать чужие классы плохо, а использовать придется

Gispoo committed:
Добавил интерфейсы, контейнер

Gispoo committed:
добавил ява доки.

sergomyaso commented on tasks/ParserCSV/src/main/java/NSU/ParserCSV/Main.java:
```diff
@@ -0,0 +1,9 @@
+    public static void main(String[] args) {
+        Container container = new Container(args);
+        CsvProcessor parser = container.getCsvProcessor();
+        parser.processCsv();
```
Круто, давай processCsv(String path)
